### ISSUE TITLE
Overloading Tor Network

### ISSUE THREAD

**Issue by silvavn** [scholarly-python-package/scholarly#131]
@ipeirotis the Tor Solution was an alternative when we had a few people using it. Right now there are about only 70 exits, and I believe the solution is not sustainable in the long term. We should really start thinking about implementing a solution where people can solve the captchas instead of just brute-forcing tor nodes.

**Comment by ipeirotis** [comment]
I think it would be a good idea.

**Comment by emilepetrone** [comment]
Is this the best solution? The reason I ask, I get many 429's but haven't hit the captcha (at least it isn't in my logs). 


```
INFO:scholarly:Exception Status code 429 while fetching page. Retrying.
INFO:scholarly:Refreshing Tor ID...
INFO:scholarly:Exception [Errno 61] Connection refused while refreshing TOR. Retrying...
INFO:scholarly:Response code 429.
```

**Comment by ipeirotis** [comment]
You have not setup Tor properly to allow for refreshing the ID. So your IP gets blocked and then you are stuck.

**Comment by stefanct** [comment]
The subject should actually be about captchas and not tor. the exact same problem exists for ordinary proxies.

I have been trying to get scholarly to work with a few proxy libraries (that scrape www.sslproxies.org and the like), e.g. https://github.com/JaredLGillespie/proxyscrape. I have modified the code so that scholarly tries another proxy automatically when it detects problems/in retries (similar to what the Tor code does). Even with that it is almost impossible to retrieve a larger amount of data.

I am not sure if you are aware of the following hack but it looks promising to me and I will rebase it and try to get it to work now: https://github.com/scholarslab/scholarly

**Comment by emilepetrone** [comment]
@stefanct interesting, its definitely an interesting little problem. 

@ipeirotis Running it again, Tor was able to refresh the ID. However I was still hitting many 429s. This is on Mac which may have something to do with it. I believe the ports are the same as Linux. 

When I look into it further, the responses are because Google is catching the requests and rejecting them. I have yet to hit the captcha. 

Is anyone else trying to run scholarly on a Mac?

**Comment by silvavn** [comment]
@stefanct actually we are aware of that "hack" @janruh had implemented it a while ago and thats where I got my repository from. 

I am on the makes to release a version using Selenium because it is more random and it gets caught by gscholar less often. 

My concern is that the purpose of Tor is not to scrape vast amounts of data and it seems like lots of people currently use this package. With that, noone can actually use it because all exits get flagged by gscholar as non-human.

So yes, Selenium is probably the solution and people will have to solve captchas if they want to collect data.

**Comment by stefanct** [comment]
I have rebased the hack on top of my PR changes in https://github.com/stefanct/scholarly/tree/selenium_captchas
I was able to retrieve quite a few pages but one gets a 403 sooner than later (that is currently not handled properly). It should also be possible to auto-continue once the captcha is solved correctly. Will look into it more tomorrow. Shall we open another issue for this?

**Comment by ipeirotis** [comment]
@stefanct Not very familiar with Selenium, but can't you refresh the ID of the Tor or proxy and continue?

I am honestly not a big fan of the idea of having to solve captcha's manually, as it invalidates a lot of projects that need to run the background. I can see some uses for people that are willing to monitor closely the crawling process.

@silvavn Not clear what you are working on in the develop branch. Could you please provide a bit of visibility on what you are working on? Perhaps it is time to merge into master and release a next minor version?

**Comment by silvavn** [comment]
@ipeirotis I am currently working on the selenium branch. 

As for the captcha discussion: IMHO, Tor is not gonna handle the amount of usage of scholarly. We just have a couple nodes for hundreds of people that use the package daily. Unfortunately we have reached a point where almost all nodes are marked as spam all the time. If we dont limit the usage rate introducing captchas all the nodes will soon be unusable.

I was a fan of using tor to brute force it as well, but again, if a few people use it, its not a problem. Now that it is public, everyone uses tor and it takes super long to find a node that works.

i wont be able to develop much for the next few weeks because I am working on my research but I can make time to quickly review code.

**Comment by ipeirotis** [comment]
@silvavn Do you use the `develop` branch for the selenium development? If not, can we scrap the current develop branch and start incorporating the submitted PRs into develop>master?

**Comment by silvavn** [comment]
I did not, you can go ahead and use develop, I have a separate branch for
selenium

Victor Silva

This message was typed and sent from a mobile phone. Please excuse any
typos.

**Comment by liujiajun** [comment]
The package is indeed overloading the Tor network.

I get 429/403 way more often than I did a few weeks ago. Now nearly every Tor exit IP is blocked by Google Scholar, making the Tor solution essentially not usable.

**Comment by liujiajun** [comment]
> @stefanct Not very familiar with Selenium, but can't you refresh the ID of the Tor or proxy and continue?
> 
> I am honestly not a big fan of the idea of having to solve captcha's manually, as it invalidates a lot of projects that need to run the background. I can see some uses for people that are willing to monitor closely the crawling process.
> 
> @silvavn Not clear what you are working on in the develop branch. Could you please provide a bit of visibility on what you are working on? Perhaps it is time to merge into master and release a next minor version?

I also prefer not to solve captchas manually. At this points of time, since Tor network is much overloaded, I think the best way to run Scholarly in background is to buy proxies. This would significantly reduce the chance of 429/403 errors.

**Comment by stefanct** [comment]
@liujiajun that's fine but an orthogonal problem/solution *and* already possible with the existing code.

@silvavn in ae453bc1 you have introduced the auto-scraping of publication bibtex data: `_scholar_pub` calls `_get_bibtex` which issues an additional web transfer for every publication in the list(!). AFAICS that's not necessary in many use cases but increases the number of transactions and thus the likelihood of getting detected by google a lot. Since that change was not long ago I think it contributed greatly to the recent problems many people face currently. I will continue to look at this in more detail tomorrow. I would like to make the additional loads optional by adding an extra parameter or even function for that (I don't have enough understanding yet to propose something more concrete). Do you have any comments regarding that?

In better news: I have a fairly well working implementation for solving captchas that is based on the hack. It supports proxies, autodetects when a captcha is needed and has been solved correctly etc. It's not perfect but makes the package useful again. I plan to release it once I get the `_get_bibtex` problem sorted out and my data retrieved... :)

**Comment by ipeirotis** [comment]
@stefanct I opened https://github.com/scholarly-python-package/scholarly/issues/137 for the multiple network calls issue. It does cause a significant amount of additional network traffic. I tried to fix it but the codebase is a bit too convoluted and not so well documented. I think that the code just needs to record the link for the bibtex generation, but only execute the network call when the "fill" function is called.

**Comment by liujiajun** [comment]
@ipeirotis Do you mean calls to URLs like this?
```https://scholar.googleusercontent.com/scholar.bib?q=info:1T2cxHtVg20J:scholar.google.com/&output=citation&scisdr=CgUI-rkwEPzwkylj4nA:AAGBfm0AAAAAXthm-nDq7UBr561uwGfA25J65QIKyak9&scisig=AAGBfm0AAAAAXthm-lmu-s-8WXtf3ya2swj6E7ONAXiW&scisf=4&ct=citation&cd=-1&hl=en```

From my experience, once I get a 429 for of of these urls, the subsequent calls to this url will always receive 429/403 even if I change my IP. I suspect this is because of this URL generated by Google Scholar varies among different users, so Google Scholar will just block this URL entirely if a bot tried to visit it.

**Comment by ipeirotis** [comment]
v.4.0 resolves the issue to a large extent.
